### PR TITLE
feat: allow full script syntax in event section

### LINKED ISSUE #65: Feature request: allow condition in events
**Is your feature request related to a problem? Please describe.**
I'd like to display a message on the plate when a button is pressed, but only when the sun is up.

**Describe the solution you'd like**
Use conditions, pretty much like in automations, in object events.

```
      - obj: "p2b2"  # light-switch toggle button warning
        event:
          "down":
            - condition: template
              value_template: "{{ (states('switch.plate_4_switch_12') == 'on' and state_attr('sun.sun','elevation') > 15) }}"
            - service: openhasp.command
              target:
                entity_id: openhasp.plate_4
              data:
                keyword: jsonl
                parameters: >-
                  {"page":2,"id":200,"obj":"obj","x":5,"y":55,"w":230,"h":160,"shadow_opa":140,"shadow_color":"black","shadow_width":20,"shadow_spread":0}
                  {"page":2,"id":201,"obj":"label","parentid":200,"x":10,"y":10,"w":215,"h":110,"text_font":24,"align":"center","text":"\uE595 It's daytime!\nPlease pull up\nthe covers instead!"}
                  {"page":2,"id":202,"obj":"btn","parentid":200,"x":20,"y":110,"w":190,"h":40,"text":"Thank you: a greener world","mode":"break","align":"center"}
```

**Describe alternatives you've considered**
Putting this in a script outside the component. But this makes the whole thing very clunky if you have lots of things like this.

**Additional context**

### PR DESCRIPTION
**Pull request by akloeckner**
This upgrades the current "list of services" restriction to allow the full syntax of HA scripts.

fixes #65

As a side-effect this also requires to generate a context for the script to run.

see #96

This can also be seen as a preparatory step to allow defining HASPObject-wide variables.

see #63

I believe, this should be even backwards-compatible, because a list of services is also a valid script.

### CODE REVIEW

**Comment by dgomes** [comment]
This is nice... but it moves from **services** to **scripts** 

Don't think this is the best path...

**Comment by akloeckner** [comment]
Isn't a "list of services" also a valid "script"? If it is, this should not break anything. Instead, it should make possible a whole range of new things, such as requested in the referenced issues. 

But I might miss something not so obvious?

(As a side-effect, we would get in line with most of the other integrations. At least I am not aware of a single integration, that allows only a subset of the script syntax as actions. Again, I might miss something.)

**Comment by dgomes** [comment]
This custom integration was based on https://www.home-assistant.io/integrations/universal/ which supports services only.

How to deal with the breaking change ?

**Comment by akloeckner** [comment]
> This custom integration was based on https://www.home-assistant.io/integrations/universal/ which supports services only.

Ah, I see. It even only admits a single service, which is not supported here, as far as I can see. (A list is required.)

> How to deal with the breaking change ?

Not sure, if this question goes to me. If it does: I still don't see anything that could break. But I still might be mistaken. Just in case, a notice could be added in the release page as was done for [0.6.0](https://github.com/HASwitchPlate/openHASP-custom-component/releases/tag/0.6.0). (If it does not: Never mind my answer. 😉)

**Comment by dgomes** [comment]
@nagyrobi since you have a big setup, can you test this PR :) ?

**Comment by akloeckner** [comment]
Hey, any news on this? I still don't see any breaking change here. Also, I have this running on my instance for half a year now, with no problems so far. The opposite: it allows me to do a whole range of things which wouldn't be possible with services alone. I'll give examples, if you like.

In my opinion: even if this were a breaking change, the opportunities outweigh the risks, by far. But that would be up to you too judge. 

For curiosity, I also checked the history, and the actual breaking change was in this commit: https://github.com/HASwitchPlate/openHASP-custom-component/commit/e31fd76e8ab452ff72af89a29975c07267749cad


Bottom line: I really think this is a useful contribution, which I would rather like to see in the upstream codebase than in my personal fork, which noone uses but me. Especially, since 0.7.0 seems to be approaching (the documentation started to prompt my to switch to it's most current version 0.7.0 lately).

**Comment by akloeckner** [comment]
Checking back... Any news here?

**Comment by nagyrobi** [comment]
> @nagyrobi since you have a big setup, can you test this PR :) ?

I'm sorry not anymore. I moved to ESPHome with all my displays meanwhile.

**Comment by nagyrobi** [comment]
@dgomes LVGL is still [work in progress](https://deploy-preview-3510--esphome.netlify.app/cookbook/lvgl#lvgl-cook) but I was able to migrate almost all functionality I previously had.

**Comment by dgomes** [comment]
like the fact that logic can be handled on the device, but coding the everything in through yaml... not the best option IMHO. JSONL approach seams more sensible to me.

**Comment by nagyrobi** [comment]
I always looked for a way to ditch MQTT from the system, now this is finally possible. Also, for our needs I need zero templates and customizations on HA side, I can do HA service calls directly from the device, no middleware required. HA seems to be faster without the ~200 templates I prevoiusly had to maintain for 6 displays.

I can have any degree of flexibility I want, from C++ lambdas even back to templates.
Conceptually ESPHome is brilliantly thought out. Compilation is very flexible, it generates smaller binaries with only the desired functionality and no Tasmota-ish data losses threatening.

**Comment by nagyrobi** [comment]
Oh and moodlight color is restored every time regardless of HA or device state as it's being stored on the device itself.

**Comment by akloeckner** [comment]
🤔  Makes me wonder if I should abandon this  PR altogether and switch as well. 😉 

Jokes aside: Before I find the time for such an adventure, I would still highly appreciate the script feature to become part of this project. And be it only for those who follow...

**Comment by dgomes** [comment]
> Oh and moodlight color is restored across reboots every time regardless of HA or device state as it's being stored on the device itself.

This could be easily added to the current firmware...

!Back_to_topic

the breaking change would be changing from single service to list of services... but if we are releasing 0.7 I guess we can accept some breaking changes @fvanroie

**Comment by akloeckner** [comment]
Just to remind: the breaking change from single service to list of services has been done way back in the past. From my I understanding, we're discussing a pure extension of functionality here, without breaking changes.

**Comment by dgomes** [comment]
I'm worried with the EVENT_SCHEMA validation change

**Comment by akloeckner** [comment]
Ah, ok. I'll check more thoroughly:

* The script schema checks for a list of script_actions:
https://github.com/home-assistant/core/blob/3c960b7d4e7a23cedb765226a98e35d1b4ab52e8/homeassistant/helpers/config_validation.py#L1294

* Where a script_action is a dictionary of any action_type_schema:
https://github.com/home-assistant/core/blob/3c960b7d4e7a23cedb765226a98e35d1b4ab52e8/homeassistant/helpers/config_validation.py#L1281

* And one of these action types is a service schema:
https://github.com/home-assistant/core/blob/3c960b7d4e7a23cedb765226a98e35d1b4ab52e8/homeassistant/helpers/config_validation.py#L1863


So, I'm confident that the current `[SERVICE_SCHEMA]` check is actually a special case of the `SCRIPT_SCHEMA` check.

Which makes sense, because a list of service calls is a perfectly valid script.

**Review by dgomes — APPROVED** [review]

**Comment by nagyrobi** [comment]
> This could be easily added to the current firmware...

Doesn't seem so, waiting for it for almost 3 years now: https://github.com/HASwitchPlate/openHASP/issues/99#issuecomment-851626353

**Comment by fvanroie** [comment]
> > This could be easily added to the current firmware...
> 
> Doesn't seem so, waiting for it for almost 3 years now: [HASwitchPlate/openHASP#99 (comment)](https://github.com/HASwitchPlate/openHASP/issues/99#issuecomment-851626353)

That issue is closed now... it doesn't seem anyone missed that feature in the past 3 years.

**Comment by fvanroie** [comment]
Is this still active?

**Comment by akloeckner** [comment]
Yes. It's running on my system since I started the PR (more or less). Without any problems.

**Comment by akloeckner** [comment]
I rebased onto main to test #122. No further changes.

**Comment by akloeckner** [comment]
> Is this still active?

I was wondering: is there anything else you need from me to proceed on this?

**Comment by fvanroie** [comment]
I have no knowledge of the HA inner workings to decide one way or the other. I will let you guys figure it out.
I was just curious since this issue has been open for some time now.

**Comment by akloeckner** [comment]
Ok, thanks for clarifying! To my understanding, Diogo had approved the changes on March 3. So, I assumed this had handed it over to you somehow. 

@dgomes, anything missing from me? Please let me know!

**Comment by dgomes** [comment]
I think this is good to go
